src/connectivities.jl

```diff
@@ -48,3 +48,27 @@ end
 function Base.show(io::IO, c::Connectivity{PL}) where {PL}
   print(io, "$PL$(c.list)")
 end
+
+"""
+Convention used to reconstruct the faces of a polytope
+from a list of vertices.
+"""
+abstract type OrderingConvention end
+
+"""
+    connectivity(polytope, convention, k)
+
+Return the list of `Connectivity` elements used to construct
+all `k`-faces of a polytope, following an ordering convention.
+"""
+function connectivity(polytope::Type{<:Polytope}, convention::OrderingConvention, k::Integer)
+  connectivity(polytope, convention, Val(k))
+end
+
+function facets(p::Polytope{N}, ordering=GMSH) where {N}
+  faces(p, N-1, ordering)
+end
+
+function faces(p::Polytope{N,Dim}, rank, ordering=GMSH) where {N,Dim}
+  (materialize(ord, p.vertices) for ord in connectivity(typeof(p), ordering, rank))
+end
```

Comment:
I am not getting how these functions interplay with the existing `connect` function. Can you clarify? Also it seems that the `facets` and `faces` functions don't belong to the connectivities.jl source file.

Reply:
`connectivity` is the function that needs to be overloaded with a custom `OrderingConvention` type. See [this line](https://github.com/JuliaGeometry/Meshes.jl/pull/15/files#diff-0c06e9261f736acae84a9cdcf8553f23a623c77979a6cadb535244fc3f6e2dfbR23) for an example.
Basically, you give a polytope, a convention and a rank, and it will build an array of `Connectivity` elements using `connect`, which correspond to the faces of the chosen rank. `faces`/`facets` add the materialization step to retrieve an array of faces.
I put these functions here because they are directly related to the notion of connectivity.

Reply:
I am not sure I am following the naming here. We have `connect` which effectively holds the connectivity list of the polytope. Your line example seems to implement the facets behavior returning a list of connectivity lists for segments? I am confused. Are you trying to generalize `facets` to any rank? The naming is really important here because we already have the concept of Connectivity and the function connect. A third function called connectivity doesn't help. It would be nice to find a better term for the concept you are trying to implement here. Also you use a keyword argument `ordering` to specify a OrderingConvention. Shouldn't this keyword argument be `convention` instead?

Reply:
I agree for the name of the keyword argument.
`faces` generalizes the retrieval of faces of any rank, just like you define it for a `Mesh`. `facets` calls `faces` with rank `N-1` for a N-polytope, and is therefore also generalized. The difference with the method for `Mesh`es, is that since the connectivity is bundled with the struct, you don't need a convention. So you only need `faces(mesh, rank)` for a mesh, and `faces(polytope, rank, convention)` for a polytope.

> Your line example seems to implement the facets behavior returning a list of connectivity lists for segments?

The GMSH convention, as you implemented it, returned only the facets of an object (it was only used to define the `facets` methods). Therefore, while a convention should allow to return any face of an object based on vertices, this convention is currently limited to returning facets.
As I understand it, a `Connectivity` instance associates a list of vertex indices with a polytope. `connectivity` describes how faces of an object are related to its vertices, returning a `Connectivity` for each face to materialize them later on (currently, `faces`/`facets` does that right after calling `connectivity`). I agree that both `Connectivity`/`connectivity` are a bit different and thus bring confusion, so we need to fix it.

Reply:
The thing is, you could just define `faces` with the logic implemented in `connectivity` so that is always returns `materialize`d faces, but having to explicitly call `materialize` in each method defined by a convention (that is, for many objects/ranks) doesn't seem right to me.